### PR TITLE
Fixes for various problems related to handeling of discrete variables.

### PR DESCRIPTION
**Pull request by hvasbath**
My proposed fixes for problems in these posts:
https://github.com/pymc-devs/pymc3/issues/655
https://github.com/pymc-devs/pymc3/issues/639

My model finally runs and is producing traces. So I tested the changes ;) .

So finally I try to use git @twiecki ;)
Feel free to comment and improve it.
Best regards.
Hannes

### CODE REVIEW

**Comment by twiecki** [comment]
@eigenblutwurst Congrats! There should be a badge for first PR submitted :).

I'll try to take a look soon.

File: pymc3/step_methods/metropolis.py [hunk 1]
```diff
@@ -109,15 +113,23 @@ def astep(self, q0):
...
         delta = self.proposal_dist() * self.scaling
         if self.any_discrete:
-            if self.all_discrete:
-                delta = round(delta, 0).astype(int)
-                q0 = q0.astype(int)
-                q = (q0 + delta).astype(int)
-            else:
-                delta[self.discrete] = round(delta[self.discrete], 0).astype(int)
-                q = q0 + delta
+            check_bnd = 1
```
Comment (fonnesbeck) [inline pymc3/step_methods/metropolis.py:119]: Can just use a bool here.

File: pymc3/step_methods/metropolis.py [hunk 2]
```diff
@@ -109,15 +113,23 @@ def astep(self, q0):
...
+            while check_bnd == 1:
```
Comment (fonnesbeck) [inline pymc3/step_methods/metropolis.py:120]: `== 1` is not necessary

File: pymc3/step_methods/metropolis.py [hunk 3]
```diff
@@ -109,15 +113,23 @@ def astep(self, q0):
...
+                if self.all_discrete:
+                    delta = round(delta, 0).astype(int)
+                    q0 = q0.astype(int)
+                    q = (q0 + delta).astype(int)
```
Comment (fonnesbeck) [inline pymc3/step_methods/metropolis.py:124]: You are already adding 2 integers, so no need to cast again. Maybe best to add then cast?

```
delta = round(delta, 0)
q = (q0 + delta).astype(int)
```

**Comment by fonnesbeck** [comment]
Would it work to transform the discrete variable to an unbounded space and propose there, before back-transforming and rounding (since we are rounding anyway)? This would avoid bounds checking at every iteration.

**Comment by hvasbath** [comment]
In my case it wouldnt work I think as my discrete variables refer to indices in a matrix and cannot really be changed. But I so far do not really understand that transformation stuff you have been implementing there. So if you can think of sth. I would be happy to remove that bound checking as it is now.
Thanks for the other comments I will include them soon.

**Comment by fonnesbeck** [comment]
The transformations allow proposals to take place on an unbounded space, so that bounds checking is not required. In your case, however, it sounds like the variable is categorical, and not merely discrete. Probably best to have a proposal that just draws from the prior (e.g. a discrete uniform) for categorical variables.

**Comment by twiecki** [comment]
Is this solved? Can we close this?

**Comment by hvasbath** [comment]
I guess so. Did anyone test the branch that you merged? I recently only used this branch here.

**Comment by twiecki** [comment]
@eigenblutwurst Do you mind checking if the problem is fixed for you on master? I'm happy to merge this if it solves your problem.

**Comment by hvasbath** [comment]
The problem is fixed on master. The DiscreteBoundcheck that is implemented here otherwise is probably not what you intend to merge with the master I can understand. If you do want to merge that I will need to commit another change I made to that first, there was a small problem. So probably rather close it?

**Comment by twiecki** [comment]
Thanks for checking @eigenblutwurst, I agree.

**Pull request closed without merging**